### ISSUE TITLE
How does this interact / conflict with customising rails logging?

### ISSUE THREAD

**Issue by nruth** [heroku/rails_stdout_logging#16]
For instance, I want to prefix Rails.logger.error messages with "error" so I can find them in papertrail.

I've been trying a number of gems and not really found anything that fits, but it's made me wonder what they're doing to the logging set up here, and if there are load-order issues.

For instance lograge, shog, and concise_logging. 

It'd be handy if the readme could give some general advice on this.

**Comment by schneems** [comment]
I don't know, it's not come up before. I would guess that you could use a railtie to extend the current `config.logger` using an `after_initialize` block, but i've not tried it. If you're only after exception information, you might want to look at how the honeybadger client collects exceptions. You could use a similar technique to catch the exceptions, re-format the messages and then output them to the logs.

**Comment by nruth** [comment]
after_initialize sounds like the trick I've missed. Thank you!

I'm more after having extra info in the logs to help find things in papertrail when searching (e.g. search for everything tagged error to see the error stream, even if they came from us using Rails.logger.error rather than from exceptions).

I guess the issue is more what does this gem do to the stock rails logger and formatting, and assumptions other gems might be making about extending or replacing it. Something that doesn't help is I'm conflating changing how rails uses the logger (lograge) with what the logger actually outputs. 

I'll try and come back to this after some sleep and make a PR for the readme if I learn anything that might be helpful.

**Comment by nakhli** [comment]
I have the same need as @nruth : I would like to change the stdout logger formatting to add severity as a  prefix to log lines. 

@schneems is there an easy way to have `rails_stdout_logging` use my custom log formatter. it is ignoring the one I set in `production.rb` using `config.log_fromatter` ...

**Comment by nruth** [comment]
I didn't find a solution in the summer, but have made some progress just now. I believe I've isolated the problem to this gem's logger set up in https://github.com/heroku/rails_stdout_logging/blob/master/lib/rails_stdout_logging/rails.rb because it doesn't set a formatter (so prints the raw message).

I've put how I came to this conclusion as a footnote in case it's interesting or someone wants to check the working out.

Changing this gem's rails.rb to the following puts formatting back into the rails production environment. I haven't tried it on Heroku but it should be the same.
```ruby
    def self.heroku_stdout_logger
      logger       = Logger.new(STDOUT)
      logger.formatter = Logger::Formatter.new
      logger       = ActiveSupport::TaggedLogging.new(logger) if defined?(ActiveSupport::TaggedLogging)
      logger.level = Logger.const_get(log_level)
      logger
    end
```

The added line is `logger.formatter = Logger::Formatter.new`.
I'm not sure why this isn't necessary in development, but is in production; likely something different rails does for development logging.

A better solution would keep the application-specified log_formatter, e.g. `logger.formatter = Rails.application.config.log_formatter` or `Rails.application.config.log_formatter.class.new` but I don't know if that'll work at this stage of booting the rails app. 

Going forward, it looks like a patch is needed to stop this gem uninstalling rails' log_formatter.

The trouble is, this will change the log output of existing apps (to the verbose Logger::Formatter format). People may be relying on the log format to do log mining etc. They could change the logging in their rails app by commenting out `config.log_formatter = ::Logger::Formatter.new` in production.rb, but need to actually realise it's happening.

Semantically this seems like a major version change, but as an injected rather than installed gem, that approach doesn't really work?

A work-around could be added to the readme, perhaps installing a replacement logger using the hook schneems mentioned, with this gem providing a factory method for a correctly configured rails STDOUT logger, with a user-specified formatter as a method param.

---

I did some experimenting with various combinations of rails environment and gems on my local machine, and believe I've isolated the problem to rails_stdout_logging gem. 

In development env, with rails_stdout_logging enabled and rails log_formatter set to Logger::Formatter in development.rb config, everything is formatted correctly. Each line is printed twice though.

By formatted correctly I mean e.g. `Rails.logger.tagged('payment') { Rails.logger.error 'hello' }` outputs `E, [2015-12-24T14:12:01.289614 #26773] ERROR -- : [payment] hello` and not only `[payment] hello`.

In production env things get weird. The logfile is formatted correctly, but unlike in development, stdout receives a single line, and it's printed raw instead of formatted by Logger::Formatter.

- I ruled out lograge as the source of the problem, as removing it didn't help and its own lines are formatted correctly when the formatting is working
- I narrowed it down to this gem by trying without it, and trying with only this and not the rest of 12 factor rails.

**Comment by nakhli** [comment]
@nruth thanks for sharing your experience. The method you metioned `self.heroku_stdout_logger` is called in a `config.before_intialize` context (see [source](https://github.com/heroku/rails_stdout_logging/blob/master/lib/rails_stdout_logging/railtie.rb#L5)). I was expecting the set logger to be set before initialization and thus I would be able to override it in my `production.rb` config. But it is not the case. In heroku, the first line of log is done using my new logger then my logger is discarded :sob:.

**Comment by nruth** [comment]
TLDR: I don't think we can cleanly replace the logger because other rails components have made references to it, and we'd also need to update all of them (without knowing what they are), or have this gem use `initialize_logger` instead of `bootstrap_hook`?

---

I'm not familiar with the Rails boot sequence and am getting in over my head a bit, but think I've found some insight in the configuration and initialization docs.

http://guides.rubyonrails.org/initialization.html
http://guides.rubyonrails.org/configuring.html#initializers

From this, I believe the config is evaluated first (perhaps in [load_environment_hook](http://guides.rubyonrails.org/configuring.html#initialization-events)), then the before_initialize block is run in the `bootstrap_hook` event to install this logger over the top.

Some rails frameworks copy / reference this logger later during initialization, and I assume that changing Rails.logger after initialization does not automatically update all frameworks that made references to it, leaving us with half-broken logging unless we also update all the other frameworks' loggers?

> action_controller.logger Sets ActionController::Base.logger - if it's not already set - to Rails.logger.

> active_record.logger Sets ActiveRecord::Base.logger - if it's not already set - to Rails.logger.

> action_mailer.logger Sets ActionMailer::Base.logger - if it's not already set - to Rails.logger.

etc…

I don't know if this list of loggers to re-assign would only be rails components, or also need to include gems brought in by the user application.

It seems (without testing it) that because this gem uses the `bootstrap_hook` event to insert the logger we may not have a clean way to override it, without also overriding it in all frameworks that make references to Rails.logger during initialization. 

We could hook into something like `i18n.callbacks` or `active_support.deprecation_behavior`, which happen before anything references the logger, but it seems naughty since it's nothing to do with logging or any kind of public API, and assumes the documented order is correct, no undocumented logger references are made, and that these events will not change or change order (which [according to stackoverflow](http://stackoverflow.com/a/8673034/395180) 
they did in Rails 3 -> 4 )

`initialize_logger` might be a better event for this gem to hook into, so 
that we can use `bootstrap_hook` (before_initialize blocks) to override it.

> initialize_logger Initializes the logger (an ActiveSupport::Logger object) for the application and makes it accessible at Rails.logger, provided that no initializer inserted before this point has defined Rails.logger.

**Comment by schneems** [comment]
I think #21 can help here but we need help getting it over the line, currently tests are failing.

**Comment by hoserdude** [comment]
Given the spirit of this gem (to ensure that output goes to STDOUT), the correct implementation is to honor existing Loggers//TaggedLoggers that have `IO::<STDOUT>` as their `logdev`.  If none are configured, override as needed.  This should really be a pre-flight safety check for heroku deployments, not an indiscriminate override. 

Having said that, I don't have a solution up my sleeve as of yet (just hit this issue), but I'll fork and see what I can do.

**Comment by schneems** [comment]
My goal is to make the gem not needed. Rails 5 has several supports baked in that makes this gem irrelevant. 

This is one such feature that helps: https://github.com/rails/rails/blob/31f6bb23929b1fafdc3390b85113dcbcd4b3203a/activesupport/lib/active_support/logger.rb#L10-L19

**Comment by artemv** [comment]
Btw I noticed a solution to this: https://gist.github.com/juno/1aea0618ca298fa69bf8
Didn't try it myself (I just removed rails_12factor, not a big deal), but it says it's compatible.

**Comment by hoserdude** [comment]
I ended up using the Railtie approach as well, but mine was a bit more complex because I wanted my own TaggedLogger implementation as well as LogFormatter.  In the end I got the JSON formatted output I wanted (and search has improved a lot with LogEntries).

**Comment by schneems** [comment]
Sorry for the extremely delayed response. If anyone has some documentation they would like to add on how to make this easier please let me know. I'm going to close this issue for now.
